src/internet/radiobrowser/radiobrowserservice.cpp

```diff
@@ -0,0 +1,390 @@
+  reply->deleteLater();
+  QJsonDocument document = ParseJsonReply(reply);
+
+  StreamList list;
+  QJsonArray contents = document.array();
+  qLog(Debug) << "RadioBrowser station list found:" << contents.size();
+  for (const QJsonValue& c : contents) {
+    QJsonObject item = c.toObject();
+    ReadStation(item, &list);
+  }
+
+  PopulateStreams(item, list);
+}
+
+void RadioBrowserService::ReadStation(QJsonObject& item, StreamList* ret) {
+  Stream stream;
+  stream.name_ = item["name"].toString();
+  QUrl url(item["stationuuid"].toString());
+  url.setScheme(url_handler_->scheme());
+  stream.url_ = url;
+  QUrl favicon(item["favicon"].toString());
+  stream.favicon_ = favicon;
+  ret->append(stream);
+}
+
+Song RadioBrowserService::Stream::ToSong(const QString& prefix) const {
+  QString song_title = name_.trimmed();
+  if (!song_title.startsWith(prefix)) {
+    song_title = prefix + " " + song_title;
+  }
+
+  Song ret;
+  ret.set_valid(true);
+  ret.set_title(song_title);
+  ret.set_artist(name_);
+  ret.set_url(url_);
+  ret.set_art_automatic(favicon_.toString());
+  return ret;
+}
+
+void RadioBrowserService::Homepage() {
+  QDesktopServices::openUrl(homepage_url_);
+}
+
+PlaylistItem::Options RadioBrowserService::playlistitem_options() const {
+  return PlaylistItem::PauseDisabled | PlaylistItem::SeekDisabled;
+}
+
+void RadioBrowserService::PopulateCategory(QStandardItem* parentItem,
+                                           QStringList& elements) {
+  if (parentItem->hasChildren())
+    parentItem->removeRows(0, parentItem->rowCount());
+
+  for (const QString& element : elements) {
+    QStandardItem* item = new QStandardItem(
+        IconLoader::Load("icon_radio", IconLoader::Lastfm), QString());
+    item->setText(element);
+    item->setData(RadioBrowserService::Type_CategoryItem,
+                  InternetModel::Role_Type);
+    item->setData(true, InternetModel::Role_CanLazyLoad);
+    parentItem->appendRow(item);
+  }
+}
+
+void RadioBrowserService::PopulateStreams(QStandardItem* parentItem,
+                                          StreamList& streams) {
+  if (parentItem->hasChildren())
+    parentItem->removeRows(0, parentItem->rowCount());
+
+  for (const Stream& stream : streams) {
+    QStandardItem* item = new QStandardItem(
+        IconLoader::Load("icon_radio", IconLoader::Lastfm), QString());
+    item->setText(stream.name_);
+    item->setData(QVariant::fromValue(stream.ToSong(name_)),
+                  InternetModel::Role_SongMetadata);
+    item->setData(InternetModel::PlayBehaviour_SingleItem,
+                  InternetModel::Role_PlayBehaviour);
+
+    parentItem->appendRow(item);
+  }
+}
+
+QDataStream& operator<<(QDataStream& out,
+                        const RadioBrowserService::Stream& stream) {
+  out << stream.name_ << stream.url_;
+  return out;
+}
+
+QDataStream& operator>>(QDataStream& in, RadioBrowserService::Stream& stream) {
+  in >> stream.name_ >> stream.url_;
+  return in;
+}
```

Comment:
I don't think these operators or the `qRegisterMetaTypeStreamOperators` should be needed unless you're writing them to settings. Some of the services use `CachedList`, which does this.

Reply:
Removed, thank you